src/Repository/DefaultUnleashProxyRepository.php

```diff
@@ -0,0 +1,207 @@
+
+        foreach ($this->configuration->getHeaders() as $name => $value) {
+            $request = $request->withHeader($name, $value);
+        }
+
+        $apiKey = $this->configuration->getProxyKey();
+        if ($apiKey === null) {
+            // The only way we can get here is if this is manually constructed without the builder
+            // @codeCoverageIgnoreStart
+            throw new LogicException('No api proxy key was specified');
+            // @codeCoverageIgnoreEnd
+        }
+
+        $request = $request->withHeader('Authorization', $apiKey);
+
+        $response = $this->httpClient->sendRequest($request);
+        $body = json_decode($response->getBody(), true);
+        if ($body === null) {
+            return null;
+        }
+        if (is_array($body)) {
+            $featureData = $this->validateResponse($body);
+
+            if ($featureData !== null) {
+                if ($this->configuration->getCache() !== null) {
+                    $this->configuration->getCache()->set($cacheKey, $featureData);
+                }
+
+                return new DefaultProxyFeature($featureData);
+            }
+        }
+
+        return null;
+    }
+
+    private function getCacheKey(string $featureName, ?Context $context): string
+    {
+        if ($context === null) {
+            return $featureName;
+        }
+        $contextHash = md5($this->contextToQueryString($context));
+
+        return $featureName . '-' . $contextHash;
+    }
+
+    private function contextToQueryString(Context $context): string
+    {
+        $query = [];
+        $values = [
+            'userId' => $context->getCurrentUserId(),
+            'sessionId' => $context->getSessionId(),
+            'remoteAddress' => $context->getIpAddress(),
+        ];
+
+        $values = array_filter($values, fn (?string $value) => $value !== null);
+        $properties = array_filter($context->getCustomProperties(), fn (?string $value) => $value !== null);
+
+        foreach ($values as $key => $value) {
+            $query[$key] = $value;
+        }
+
+        $query['properties'] = [];
+
+        foreach ($properties as $key => $value) {
+            $query['properties'][$key] = $value;
+        }
+
+        return urldecode(http_build_query($query));
+    }
+
+    private function addQuery(string $url, string $query): string
+    {
+        $urlParts = parse_url($url);
+
+        // The base url is already validated, so this should only happen if a feature toggle has a name
+        // that can't form part of a URL but this shouldn't ever happen, given that this is validated
+        // by the Unleash server and UI
+
+        // @codeCoverageIgnoreStart
+        if ($urlParts === false) {
+            throw new \InvalidArgumentException('Invalid URL provided');
+        }
+        // @codeCoverageIgnoreEnd
+
+        if (!isset($urlParts['query'])) {
+            $urlParts['query'] = $query;
+        }
```

Comment:
What happens if `$urlParts['query']` is set? Are we not merging `$query` with `$urlParts['query']`? If this shouldn't happen maybe we can add a condition to the if above which is checking `if ($urlParts === false)`: `if ($urlParts === false || isset($urlParts['query'])`

Reply:
Then something super odd is going on and user has found a creative and interesting way to misuse this code. This url is built up by this code based on a root url that's passed in at startup. I'm pretty sure this should never be an issue